File: LeetCode/Practice_for_interview/Search/DFS/695.py
class Solution:

    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        def dfs(x, y):
            if x < 0 or x >= m or y < 0 or y >= n or grid[x][y] == 0:
                return 0
            ans = 1
            grid[x][y] = 0
            for xi, yi in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                ans += dfs(x+xi, y+yi)
            return ans
        
        max_area = 0
        for i in range(m):
            for j in range(n):
                max_area = max(max_area, dfs(i, j))
        
        return max_area

File: LeetCode/Practice_for_interview/Search/DFS/695.cpp
class Solution {
public:
    vector<pair<int, int>> dir = {{0,1}, {0,-1}, {1,0}, {-1, 0}};
    int dfs(vector<vector<int>> &g, int x, int y) {
        if (x < 0 || x >= g.size() || y < 0 || y >= g[0].size() ||g[x][y] == 0) return 0;
        int ans = 1;
        g[x][y] = 0;
        for (auto d: dir) {
            ans += dfs(g, x+d.first, y+d.second);    
        }
        return ans;


    }
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int ans = 0;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                int t = dfs(grid, i, j);
                ans = max(ans, t);
            }
        }
        return ans;
    }
};